### ISSUE TITLE
Question: using with 3rd party libraries that expose CompletableFuture<T> in their API

### ISSUE THREAD

**Issue by bvanderveen** [stefan-zobel/streamsupport#8]
Hi, 

Here is my scenario.

Library M (my plain vanilla Java code) depends on JAR N (3rd party open source JAR from e.g., maven). I have Project A which is an Android app that has to support some old Android API versions; and Project B which is a Java 8 server backend deployed to infrastructure which I control. Both depend on Library M.

JAR N exports the following type:

```
import java.util.concurrent.CompletableFuture;

public interface LibraryN {
CompletableFuture<String> getTheString();
}
```

Can you confirm that this `streamsupport` project does not help me build my Android app with its dependency on LibraryN, because one of my dependencies (specifically Library N) compiles against `java.util.concurrent.*` and that it is not possible for me to 'replace' the `java.util.concurrent.*` reference with a reference to the classes provided by this project?

If not, it seems worth adding a caveat to the readme along the lines of "This library cannot help you if one of your project's dependencies depends on Java 8". Maybe this is obvious and maybe I'm kinda dumb, but I can't imagine I'm the only one who will ever make this mistake. 

Otherwise—is there a way I can make my scenario work? 

Thank you for your time and consideration.

**Comment by stefan-zobel** [comment]
Hi,

I'm sorry, there is no magic other than bytecode rewriting that can make the j.u.c.CompletableFuture in Library N disappear.

streamsupport is only useful if you control the source code. `ProGuard` can be used to replace usages of the stream API in a Java 8 Jar by the streamsupport implementation, see https://sourceforge.net/p/proguard/discussion/182455/thread/8718d2cea2/

However, I doubt that it can replace CompletableFuture. You could ask @netomi (Thomas Neidhart) who is the developer - I didn't look too deeply into that ProGuard feature.
See also https://github.com/stefan-zobel/streamsupport/issues/5

Other than that the only way I see is to compile and use your own adapted version of library N.

HTH,
Stefan

**Comment by netomi** [comment]
Hi,

there seems to be no counterpart for j.u.c.CompletableFuture in the streamsupport library so its not replaced. If there would be a counterpart and ProGuard could it replace though your use-case would be perfectly supported. I am not fully aware why this class is missing, maybe Stefan can give some insights.

**Comment by stefan-zobel** [comment]
@netomi 

https://github.com/stefan-zobel/streamsupport/blob/master/src/cfuture/java/java8/util/concurrent/CompletableFuture.java

It's in the `streamsupport-cfuture` jar : https://search.maven.org/search?q=a:streamsupport-cfuture

Greetings,
Stefan

**Comment by netomi** [comment]
ah thanks, the current implementation focuses on the core streamsupport library, but we can add support for the optional modules like cfuture as well.

**Comment by bvanderveen** [comment]
@stefan-zobel @netomi Gents, thank you for weighing in.

At the end of a late night, I came to the same conclusions described in this thread.

 - `streamsupport` alone is only useful if you control the source code all the way down (or are willing to modify and recompile open source deps yourself)
 - ProGuard 6.1.0beta2 can replace references to the the default JVM implementation with references to `streamsupport` and does so automatically when it detects a dependency on `streamsupport` and a `-target 1.7` flag.
 - ProGuard does 6.2.0beta2 does not support replacing `CompletableFuture`
 - `CompletableFuture` is not in the main `streamsupport` JAR but rather in the `streamsupport-cfuture` jar.

@netomi I, for one, would love to have CompletableFuture support in ProGuard so that I can transparently use it in my Android app.

*Workaround*: For now, I have worked around by replacing references in my dependency's source code (as well as in my Android app and my JVM backend service) to `java.util.concurrent.CompletableFuture` by `java8.util.concurrent.CompletableFuture`, and recompiling everything all the way up.

@stefan-zobel Speaking as a hasty pragmatist, my two cents is that I would have benefitted from this caveat being a little more obvious in the README for streamsupport. When I read 'backport' I assumed binary compatibility. Would you be open to accepting a PR for the readme?

**Comment by stefan-zobel** [comment]
Hi Benjamin,

as to the README, what's your suggestion?

TBH, to me that "binary compatibility" expectations are always astonishing. I'm possibly a bit old school, but

 - could you plant your own `java.` package into a VM without fiddling with the boot classloader?
 - would you expect Java 6 Collections to suddenly have a `stream()` method without even touching their implementation?
 - do you anticipate to be able to use default / static methods in a Java 6 interface by simply adding a Jar to your project?

I could go on with this list, but the answer is always an obvious (to me) **"no"**.
Btw., there is already a [Readme.txt](https://github.com/stefan-zobel/streamsupport/blob/master/Readme.txt) in the project that perhaps makes this all a bit clearer. Did you see that?

And no, please let's not go through a complete PR for a single sentence that might have to be added to the README.md - a casual discussion of your proposal is sufficient.
I'll make that change when you come up with a formulation that adds something useful & doesn't treat experienced Java devs as they were nescient noobs.

Regards,
Stefan

**Comment by stefan-zobel** [comment]
Oh, and btw., ProGuard 6.1 is not beta anymore. The final 6.1 has been released less than 2 weeks ago: https://sourceforge.net/projects/proguard/files/proguard/6.1/

**Comment by bvanderveen** [comment]
Hi Stefan,

You're right on all counts, and it all makes sense and indeed should have been obvious. I'll be the first to admin that I'm a hasty and ham-fisted cretin who doesn't think things through before acting, and that I deserve all of the pain I've wrought upon myself.

I did see readme.txt, but only after I had deduced in a head-vs-wall manner everything that it says. 

Perhaps an appropriate update to the readme.md, which is all I saw in my drive-by skimming of the Github page would be something along the lines of:

> It is possible—although not supported directly by the maintainer of this project—to use ProGuard to rewrite binaries previously-compiled against Java 8 to use this implementation of Java Streams. This can be useful if you would like to use a library that depends on Java 8 Streams on Android, for example. For more information, see the [ProGuard documentation](https://www.guardsquare.com/en/products/proguard/manual/gradleplugin).

All of these issues seem to have stemmed from the same sort of misunderstanding

- https://github.com/stefan-zobel/streamsupport/issues/5 - Dude is also trying to use a java8 lib on Android, opened an issue because he was lost on how to do so.
- https://github.com/stefan-zobel/streamsupport/issues/6 - Dude didn't pick up on the fact that the whole thing is in `java8.` — a simple usage example showing `import java8.…` on the readme could have helped. As for me, I also didn't realize the whole thing was in `java8.` until late in the process.
- https://github.com/retrostreams/android-retrostreams/issues/4 - Dude also had expectations of bytecode switching based on the word 'backport'

I beseech you, kind and understanding sir, to drop a crumb in a readme.md somewhere that might save some of us poor, dumb sods developing for Android a few cycles thrashing in the darkness of our ignorance.

**Comment by stefan-zobel** [comment]
Hi Benjamin,

Point taken. I agree, a reference to ProGuard and the bytecode rewriting scenario is indeed useful.
Thanks for the suggestion, I'll add that in the next few days (1.7.1 should be out soon).

Regards,
Stefan

**Comment by bvanderveen** [comment]
Stefan,

Much appreciated. Cheers and thanks for your work on this awesome library.

Benjamin

**Comment by netomi** [comment]
@bvanderveen just for your info: we have created an internal ticket to support cfuture and other optional components of the streamsupport library as well. Adding this should not be difficult so it will hopefully be available soon.

**Comment by bvanderveen** [comment]
@netomi That's great! Glad this discussion could catalyse that support. I'll keep an eye on the latest Proguard releases. Thank you for your work!

**Comment by stefan-zobel** [comment]
@bvanderveen Done that

@netomi Hi Thomas,

thanks for looking into that.

I noticed that the ProGuard documentation still mentions the "streamsupport-literal" (JEP 269, Collections factory methods) component. Please forget that. That code has been merged into the core streamsupport jar in release 1.6.0. I don't maintain / update "streamsupport-literal" as a separate component anymore.

According to the Maven Central download numbers support for CompletableFuture certainly would have its user base.

Last month's downloads in order of decreasing popularity:

 - streamsupport
 - streamsupport-cfuture 
 - streamsupport-flow
 - streamsupport-atomic
 - streamsupport-literal

I don't think supporting "atomic" is that important. It seems to be rarely used and the performance (on Android devices) is abysmal.
Unsure about "flow" (JEP 266, reactive streams) though. I don't see a lot of downloads or even usage in standard Java (yet).

Another thing to consider is that both JEP 266 and JEP 269 are Java 9+ APIs. CompletableFuture has its own share of post-Java 8 additions of course.

HTH,
Stefan

**Comment by netomi** [comment]
@stefan-zobel thanks for the pointers, we will update the code / documentation accordingly. 
We have also a way to define what the target Java runtime shall be (the -target option), this allows to decide what parts shall be backported, I will check for all added types / methods in which JDK the are available.

**Comment by stefan-zobel** [comment]
@netomi  It's probably easiest just to search the code for  the Javadoc ""at"since 9" up to ""at"since 12" tags. To be honest, I couldn't tell in any other way at the moment where additions have been made.
